File: python/protein-translation/protein_translation.py
# Codon                 | Protein
# :---                  | :---
# AUG                   | Methionine
# UUU, UUC              | Phenylalanine
# UUA, UUG              | Leucine
# UCU, UCC, UCA, UCG    | Serine
# UAU, UAC              | Tyrosine
# UGU, UGC              | Cysteine
# UGG                   | Tryptophan
# UAA, UAG, UGA         | STOP

CODON_TO_PROTEIN = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan"
}

STOP_CODONS = set({"UAA", "UAG", "UGA"})
CODON_LENGTH = 3


def proteins(strand):
    polypeptide = []
    codons = split_into_codons(strand)

    for codon in codons:
        if (codon in STOP_CODONS):
            return polypeptide
        else:
            polypeptide.append(CODON_TO_PROTEIN[codon])

    return polypeptide


def split_into_codons(strand):
    return [strand[i:i + CODON_LENGTH] for i in range(0, len(strand),
                                                      CODON_LENGTH)]

File: go/protein-translation/protein_translation.go
package protein

import (
	"errors"
)

// ErrStop represents a stop codon
var ErrStop error = errors.New("stop codon")

// ErrInvalidBase represents an invalid base that cannot me mapped to an amino acid.
var ErrInvalidBase error = errors.New("invalid base")

var stopCodons = map[string]bool{
	"UAA": true,
	"UAG": true,
	"UGA": true,
}

var codonToProtein = map[string]string{
	"AUG": "Methionine",
	"UUU": "Phenylalanine",
	"UUC": "Phenylalanine",
	"UUA": "Leucine",
	"UUG": "Leucine",
	"UCU": "Serine",
	"UCC": "Serine",
	"UCA": "Serine",
	"UCG": "Serine",
	"UAU": "Tyrosine",
	"UAC": "Tyrosine",
	"UGU": "Cysteine",
	"UGC": "Cysteine",
	"UGG": "Tryptophan",
}

func FromCodon(codon string) (protein string, e error) {
	if _, ok := stopCodons[codon]; ok {
		return "", ErrStop
	}
	if _, ok := codonToProtein[codon]; !ok {
		return "", ErrInvalidBase
	}
	return codonToProtein[codon], nil
}

const CODON_LENGTH = 3

func FromRNA(codons string) (proteins []string, e error) {
	for i := 0; i <= len(codons)-CODON_LENGTH; i += CODON_LENGTH {
		codon := codons[i : i+CODON_LENGTH]
		codon, err := FromCodon(codon)
		if errors.Is(err, ErrStop) {
			return proteins, nil
		}
		if errors.Is(err, ErrInvalidBase) {
			return proteins, ErrInvalidBase
		}
		proteins = append(proteins, codon)
	}
	return proteins, nil
}
